### PR TITLE
Allow default values by channels

### PR DESCRIPTION
**Pull request by etienne-monsieurbiz**
### Features

- Add a new node `default_values_for_channels` in the settings yaml to define `default_values` in a channel scope.
- Display the default value per channel in admin if none is saved to DB
- Update `getCurrentValue()` to take default channel values into account

Admin:
Capture d’écran 2023-07-21 à 12 05 36

Front:
Capture d’écran 2023-07-21 à 12 06 07

No values in DB:
Capture d’écran 2023-07-21 à 12 06 49

### CODE REVIEW

File: src/Factory/Form/MainSettingsFormTypeFactory.php [hunk 1]
```diff
@@ -68,12 +68,21 @@ private function getInitialFormData(SettingsInterface $settings): array
...
         /** @var ChannelInterface $channel */
         foreach ($this->channelRepository->findAll() as $channel) {
-            $data['channel-' . $channel->getId() . '-' . Settings::DEFAULT_KEY] = $settings->getSettingsValuesByChannelAndLocale($channel);
+            $channelKey = 'channel-' . $channel->getId() . '-' . Settings::DEFAULT_KEY;
+            $data[$channelKey] = $settings->getSettingsValuesByChannelAndLocale($channel);
+            $defaultDataByChannel = $this->getDefaultValuesByChannel($settings, $channel);
+
+            if (isset($defaultDataByChannel[$channelKey])) {
+                $data[$channelKey] += $defaultDataByChannel[$channelKey];
+            }
```
Comment (delyriand) [inline src/Factory/Form/MainSettingsFormTypeFactory.php:85]: **❔ question:** ‏why the defaults value per channel are added here but not the global defaults?

**Review by delyriand — CHANGES_REQUESTED** [review]

File: src/Factory/Form/MainSettingsFormTypeFactory.php (hunk 1 above)
Comment (etienne-monsieurbiz) [reply src/Factory/Form/MainSettingsFormTypeFactory.php:85]: because we are in getChannelInitialFormData() the global default value is handle before the call to this method

File: src/Factory/Form/MainSettingsFormTypeFactory.php [hunk 2]
```diff
@@ -68,12 +68,21 @@ private function getInitialFormData(SettingsInterface $settings): array
         return $data + $this->getChannelInitialFormData($settings);
     }
 
+    /**
+     * @SuppressWarnings(PHPMD.CyclomaticComplexity)
+     */
```
Comment (jacquesbh) [inline src/Factory/Form/MainSettingsFormTypeFactory.php:73]: **🎯 suggestion:** ‏Maybe we could split the method in two?

File: src/Factory/Form/MainSettingsFormTypeFactory.php [hunk 3]
```diff
@@ -68,12 +68,21 @@ private function getInitialFormData(SettingsInterface $settings): array
...
         $data = [];
         /** @var ChannelInterface $channel */
         foreach ($this->channelRepository->findAll() as $channel) {
-            $data['channel-' . $channel->getId() . '-' . Settings::DEFAULT_KEY] = $settings->getSettingsValuesByChannelAndLocale($channel);
+            $channelKey = 'channel-' . $channel->getId() . '-' . Settings::DEFAULT_KEY;
+            $data[$channelKey] = $settings->getSettingsValuesByChannelAndLocale($channel);
+            $defaultDataByChannel = $this->getDefaultValuesByChannel($settings, $channel);
+
+            if (isset($defaultDataByChannel[$channelKey])) {
+                $data[$channelKey] += $defaultDataByChannel[$channelKey];
```
Comment (jacquesbh) [inline src/Factory/Form/MainSettingsFormTypeFactory.php:84]: **🎯 suggestion:** ‏Are you sure about this?

What about this instead?

```suggestion
                $data[$channelKey][] = $defaultDataByChannel[$channelKey];
```

File: src/Settings/SettingsInterface.php [hunk 4]
```diff
@@ -43,9 +43,9 @@ public function getFormClass(): string;
     /**
      * @SuppressWarnings(PHPMD.BooleanArgumentFlag)
      */
-    public function getSettingsByChannelAndLocale(?ChannelInterface $channel = null, ?string $localeCode = null, bool $withDefault = false): array;
+    public function getSettingsByChannelAndLocale(ChannelInterface $channel = null, string $localeCode = null, bool $withDefault = false): array;
 
-    public function getSettingsValuesByChannelAndLocale(?ChannelInterface $channel = null, ?string $localeCode = null): array;
+    public function getSettingsValuesByChannelAndLocale(ChannelInterface $channel = null, string $localeCode = null): array;
```
Comment (jacquesbh) [inline src/Settings/SettingsInterface.php:48]: **❔ question:** ‏Why did you remove the nullable possibility?

**Review by jacquesbh — CHANGES_REQUESTED** [review]

File: src/Factory/Form/MainSettingsFormTypeFactory.php (hunk 3 above)
Comment (etienne-monsieurbiz) [reply src/Factory/Form/MainSettingsFormTypeFactory.php:84]: Yeah, I'm sure, I want a union of the two arrays with the keys of `$data` if they exist in both array.
`$data[$channelKey][] =` will add me a new array on a numeric index which is not what I want

File: src/Settings/SettingsInterface.php (hunk 4 above)
Comment (etienne-monsieurbiz) [reply src/Settings/SettingsInterface.php:48]: oops, mistake, that was done by phpcs before I changed the rules to not have this change.
